### ISSUE TITLE
SYR and POSV very slow in pthreads build

### ISSUE THREAD

**Issue by david-cortes** [xianyi/OpenBLAS#3237]
**TL;DR:** SYR in the pthreads variant of OpenBLAS would run very slow when it is inside an OpenMP parallel region, even after setting the openblas threads to one - C++ code example near the bottom of the comments. Most of the time is spent in locking mechanisms.

Original:
**********************************
I'm experiencing a very large slow down from using POSV in a loop, using the openmp variant of openblas (haven't tested the other), version 0.3.15, with an AMD Ryzen 7 2700 CPU.

I do not have a minimally reproducible example, but to explain it in words, I am comparing low-rank matrix factorization models using the alternating least-squares method, using POSV for solving small linear systems in parallel (parallelizing with openmp), which is the bottleneck computation in this algorithm (code: https://www.github.com/david-cortes/cmfrec).

In short, the code is a package for recommender systems, which I'm timing on the MovieLens10M dataset, with everything in double precision. I've tried compiling the package with different options, and I get some very weird timings in the algorithm when linking to openblas:
* MKL, parallel: ~30s
* OpenBLAS, parallel: ~360s
* OpenBLAS, parallel with OPENBLAS_NUM_THREADS=1: ~240s
* OpenBLAS, but seemingly not for lapack (through R's build system for packages): ~60s
* Similar code using Armadillo: ~30s
* OpenBLAS, but using a conjugate gradient solver instead of POSV: ~20-60s depending on tolerance
* MKL, no parallelization: ~180s
* OpenBLAS, no parallelization from outside: ~640s
* OpenBLAS, no parallelization from outside + OPENBLAS_NUM_THREADS=1 and OMP_NUM_THREADS=1: ~220s

I also find it very weird that POSV somehow gets parallelized by openblas, as the problems are linear systems of rank 50 with one RHS.

**Comment by david-cortes** [comment]
Actually, from some experiments solving the linear systems differently, I'm now not so sure if POSV is what slows it down. Might be SYR which I am calling before POSV. Will do more investigations.

**Comment by brada4** [comment]
All eyes on your measurements ;-)
- dataset is known - but what to use as a test/repeater script? , something like `call dposv(.....)` or any other sequence?

- Is openblas replacing libRblas / libRlapack or something else? Fedora does that by default, Debian etc has it under alternative.

- Which timing function? e.g. I think [these](https://github.com/xianyi/OpenBLAS/tree/develop/benchmark/scripts/R) are correct. Also check `gc()` avoidance there by setting output before timed loop.

- Is the particular OpenBLAS library 1/ only single blas/lapack library in process 2/ OMP builD as confirmable via `ldd` 3/ perferably vendor package or something as easily repeatable

**Comment by martin-frbg** [comment]
POSV has SYRK on its call graph, which I remember you found performance problems with in the past (though IIRC these were resolved by reverting an earlier change). Nothing immediately suspicious in the SYR pathway except that it will be parallelized no matter how small the problem, which is probably not a good idea (driver is original GotoBLAS2 code employing AXPY)

**Comment by brada4** [comment]
syrk threshold shall be at least page, i.e (n^2)/2<page/sizeof
sqrt(2k) ... 45 being guaranteed pessimal
Say 47 to be obvious bad magic until properly assessed?

**Comment by david-cortes** [comment]
Well, I didn't manage to pin down the culprit, so this time I'm attaching a reproducible example, which I'm testing out in jupyter with the Cython extension installed. Modify the linking arguments accordingly:
```python
%load_ext Cython
```
```python
%%cython --compile-args=-fopenmp --link-args=-lopenblas --link-args=-fopenmp -I/usr/include/x86_64-linux-gnu/
import numpy as np
cimport numpy as np
cdef extern from *:
    """
    #include <string.h>
    #include <stdlib.h>
    #include <omp.h>
    #include "cblas.h"
    void dposv_(const char*, const int*, const int*, const double*, const int*, const double*, const int*, const int*);

    void ALS_iteration
    (
        int *Xcsr_p,
        int *Xcsr_i,
        double *Xcsr,
        double *A,
        double *B,
        int m, int k,
        int nthreads
    )
    {
        int one = 1;
        char lo = 'L';
        double *buffer = (double*)malloc(k*k*nthreads*sizeof(double));
        
        #pragma omp parallel for schedule(dynamic) num_threads(nthreads)
        for (int row = 0; row < m; row++)
        {
            int info = 0;
            double *buffer_thread = buffer + k*k*omp_get_thread_num();
            double *a_vec = A + row*k;
            memset(buffer_thread, 0, k*k*sizeof(double));
            memset(a_vec, 0, k*sizeof(double));
            
            for (int ix = Xcsr_p[row]; ix < Xcsr_p[row+1]; ix++)
                cblas_dsyr(CblasRowMajor, CblasUpper, k,
                           1., B + Xcsr_i[ix]*k, 1,
                           buffer_thread, k);
            
            for (int ix = Xcsr_p[row]; ix < Xcsr_p[row+1]; ix++)
                cblas_daxpy(k, Xcsr[ix], B + Xcsr_i[ix]*k, 1, a_vec, 1);
            
            dposv_(&lo, &k, &one,
                   buffer_thread, &k,
                   a_vec, &k,
                   &info);
        }
        
        free(buffer);
    }
    """
    void ALS_iteration(
        int *Xcsr_p,
        int *Xcsr_i,
        double *Xcsr,
        double *A,
        double *B,
        int m, int k,
        int nthreads
    )

def run_ALS_iteration(
    np.ndarray[int, ndim=1] Xcsr_p,
    np.ndarray[int, ndim=1] Xcsr_i,
    np.ndarray[double, ndim=1] Xcsr,
    np.ndarray[double, ndim=2] A,
    np.ndarray[double, ndim=2] B,
    int nthreads
):
    ALS_iteration(
        &Xcsr_p[0],
        &Xcsr_i[0],
        &Xcsr[0],
        &A[0,0],
        &B[0,0],
        A.shape[0], A.shape[1],
        nthreads
    )
```
```python
import numpy as np
from scipy.sparse import random

m = int(1e5)
n = int(1e4)
k = 50
density = 0.01
Xcsr = random(m, n, 0.01, format="csr", random_state=1)
np.random.seed(1)
A = np.zeros((m,k))
B = np.random.normal(size=(n,k))

Xcsr.indptr = Xcsr.indptr.astype(np.int32)
Xcsr.indices = Xcsr.indices.astype(np.int32)
```
```python
%%time
run_ALS_iteration(
    Xcsr.indptr,
    Xcsr.indices,
    Xcsr.data,
    A,
    B,
    16
)
```


MKL:
```
CPU times: user 17.8 s, sys: 1.01 s, total: 18.8 s
Wall time: 1.26 s
```

OpenBLAS built from source:
```
CPU times: user 1min 31s, sys: 1min 55s, total: 3min 26s
Wall time: 13.1 s
```

(The system openblas would instead throw warnings about openmp, despite being the openmp variant)

************

Just to clarify, I'm experiencing the problem in C code which is wrapped from R/Python, for which I'm setting custom linker arguments, and measuring ellapsed time (through jupyter's ``%%time`` and R's `microbenchmark`). This is in Debian, with the BLAS alternative set to OpenBLAS installed from the `main` repository (version 0.3.15 from experimental). Tried also compiling OpenBLAS from source (with `USE_OPENMP` and `BUILD_SHARED_LIBS`), but it didn't make a difference.

This is the `ldd` output for my openblas `.so` file - it includes `libgomp`:
```
david@debian:/lib/x86_64-linux-gnu$ ldd libopenblas.so.0
        linux-vdso.so.1 (0x00007fff33336000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fa874df1000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fa874dcf000)
        libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007fa874b19000)
        libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007fa874ad9000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa874914000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fa877253000)
        libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007fa8748cb000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fa8748af000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fa8748a9000)
```

This is the `ldd` output that I get in the compiled package through R's build system (code runs in ~60s):
```
david@debian:~/R/x86_64-pc-linux-gnu-library/4.0/cmfrec/libs$ ldd cmfrec.so 
        linux-vdso.so.1 (0x00007ffd5a1dc000)
        liblapack.so.3 => /lib/x86_64-linux-gnu/liblapack.so.3 (0x00007f59dfee8000)
        libblas.so.3 => /lib/x86_64-linux-gnu/libblas.so.3 (0x00007f59dfe83000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f59dfd3f000)
        libR.so => /lib/libR.so (0x00007f59df88e000)
        libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f59df84e000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f59df82c000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f59df665000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f59e0633000)
        libopenblas.so.0 => /lib/x86_64-linux-gnu/libopenblas.so.0 (0x00007f59dd377000)
        libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007f59dd0c1000)
        libreadline.so.8 => /lib/x86_64-linux-gnu/libreadline.so.8 (0x00007f59dd06a000)
        libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007f59dcfd2000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f59dcfaa000)
        libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f59dcf95000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f59dcf78000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f59dcf72000)
        libicuuc.so.67 => /lib/x86_64-linux-gnu/libicuuc.so.67 (0x00007f59dcd89000)
        libicui18n.so.67 => /lib/x86_64-linux-gnu/libicui18n.so.67 (0x00007f59dca83000)
        libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f59dca38000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f59dca1e000)
        libtinfo.so.6 => /lib/x86_64-linux-gnu/libtinfo.so.6 (0x00007f59dc9ef000)
        libicudata.so.67 => /lib/x86_64-linux-gnu/libicudata.so.67 (0x00007f59daed6000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f59dad09000)
```

This is the `ldd` output from the slowed down version (from Python's build system with `-lopenblas` manually added there) (code runs in 10 minutes with openblas, ~30s with MKL):
```
david@debian:~/Projects/cmfrec/cmfrec$ ldd wrapper_double.cpython-37m-x86_64-linux-gnu.so 
        linux-vdso.so.1 (0x00007ffd313d6000)
        libopenblas.so.0 => /lib/x86_64-linux-gnu/libopenblas.so.0 (0x00007f6ea7acb000)
        libgomp.so.1 => /home/david/anaconda3/lib/libgomp.so.1 (0x00007f6ea7a9e000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f6ea7a7c000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6ea78b7000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f6ea9f7c000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f6ea7773000)
        libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007f6ea74bd000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f6ea74b0000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f6ea74aa000)
        libquadmath.so.0 => /home/david/anaconda3/lib/libquadmath.so.0 (0x00007f6ea7470000)
        libgcc_s.so.1 => /home/david/anaconda3/lib/libgcc_s.so.1 (0x00007f6ea745c000)
```

And finally this is for the alternative Armadillo version (code runs in ~30s):
```
david@debian:~/R/x86_64-pc-linux-gnu-library/4.0/rsparse/libs$ ldd rsparse.so 
        linux-vdso.so.1 (0x00007ffdc6d83000)
        liblapack.so.3 => /lib/x86_64-linux-gnu/liblapack.so.3 (0x00007f068119f000)
        libblas.so.3 => /lib/x86_64-linux-gnu/libblas.so.3 (0x00007f068113a000)
        libR.so => /lib/libR.so (0x00007f0680c89000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f0680abc000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0680978000)
        libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f0680938000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f068091c000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0680757000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0681919000)
        libopenblas.so.0 => /lib/x86_64-linux-gnu/libopenblas.so.0 (0x00007f067e469000)
        libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007f067e1b3000)
        libreadline.so.8 => /lib/x86_64-linux-gnu/libreadline.so.8 (0x00007f067e15c000)
        libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007f067e0c4000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f067e09a000)
        libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f067e087000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f067e06a000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f067e064000)
        libicuuc.so.67 => /lib/x86_64-linux-gnu/libicuuc.so.67 (0x00007f067de7b000)
        libicui18n.so.67 => /lib/x86_64-linux-gnu/libicui18n.so.67 (0x00007f067db75000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f067db51000)
        libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f067db08000)
        libtinfo.so.6 => /lib/x86_64-linux-gnu/libtinfo.so.6 (0x00007f067dad9000)
        libicudata.so.67 => /lib/x86_64-linux-gnu/libicudata.so.67 (0x00007f067bfc0000)
```

**Comment by david-cortes** [comment]
From further investigation, this is a breakdown of CPU time taken in each chunk under OpenBLAS, measured with C++'s `std::chrono`:
* SYR loop: 195.2956768850001
* AXPY loop: 0.5242378820000001
* POSV: 9.118683672999994

Breakdown with MKL:
* SYR loop: 15.003344129
* AXPY loop: 0.8048905269999994
* POSV: 2.724766450999999

So it's both SYR and POSV, but with SYR being more guilty.

However, if I just run SYR alone outside of a larger code block, I don't see any slow down on these problem sizes.

****************
As a curiosity, there's another library (`lenskit`) which implements the same algorithm but using NumPy+SciPy+Numba:
https://github.com/lenskit/lkpy/blob/3172a703336ae7664790dd62050ced7185390624/lenskit/algorithms/als.py#L94
(class `BiasedMF` with `method="lu"` as the frontend)

With that library I also observe similar differences when fitting the model to the same data.

**Comment by david-cortes** [comment]
And BTW changing SYR to SYRK with `n=1` reduces the time spent there to ~125s instead. With MKL it also reduces the time spent there to ~8.35s.

**Comment by david-cortes** [comment]
Shorter example, SYR only, the problem seems to occur **only** when it is called inside an OMP parallel region:
```python
%load_ext Cython
```
```python
%%cython --compile-args=-fopenmp --link-args=-lopenblas --link-args=-fopenmp -I/usr/include/x86_64-linux-gnu/
import numpy as np
cimport numpy as np
cdef extern from *:
    """
    #include "cblas.h"

    void run_code
    (
        int k,
        double *X_in,
        double *X_out,
        int *indices, int n,
        int nthreads
    )
    {
        #pragma omp parallel for schedule(static) num_threads(nthreads)
        for (int ix = 0; ix < n; ix++)
        {
            for (int repeat = 0; repeat < 1000; repeat++)
                cblas_dsyr(CblasRowMajor, CblasUpper, k,
                           1., X_in + indices[ix]*k, 1,
                           X_out + indices[ix]*k*k, k);
        }
    }
    """
    void run_code(
        int k,
        double *X_in,
        double *X_out,
        int *indices, int n,
        int nthreads
    )

def run_c_code(np.ndarray[double, ndim=2] X_in,  ### m-by-k
               np.ndarray[double, ndim=3] X_out, ### m-by-k-by-k
               np.ndarray[int, ndim=1] indices,  ### n
               int nthreads):
    cdef int k = X_in.shape[1]
    run_code(
        k,
        &X_in[0,0],
        &X_out[0,0,0],
        &indices[0], indices.shape[0],
        nthreads
    )
    return None
```
```python
import numpy as np

k = 50
m = 10000
np.random.seed(1)
X_in = np.random.normal(size=(m, k)).astype(np.float64)
X_out = np.zeros((m, k, k)).astype(np.float64)
indices = np.arange(m).astype(np.int32)
np.random.shuffle(indices)
```
```python
%%time
run_c_code(X_in, X_out, indices, 16)
```

With OpenBLAS:
```
CPU times: user 1min 13s, sys: 1min 52s, total: 3min 5s
Wall time: 11.8 s
```

With OpenBLAS setting `OPENBLAS_NUM_THREADS` and `OMP_NUM_THREADS` to 1:
```
CPU times: user 19 s, sys: 1min 36s, total: 1min 55s
Wall time: 7.95 s
```

With OpenBLAS passing `nthreads=1` but without setting the environment variables:
```
CPU times: user 33.5 s, sys: 22.2 s, total: 55.7 s
Wall time: 13.9 s
```

With MKL:
```
CPU times: user 10.5 s, sys: 91.9 ms, total: 10.6 s
Wall time: 682 ms
```


When running single threaded:
* OpenBLAS: ~7s
* MKL: ~4.7s

**Comment by martin-frbg** [comment]
POSV in OpenBLAS is same as in Reference-LAPACK, so the difference will be entirely in the BLAS functions called within. Of there, GEMV has already drawn criticism for being inefficient at small problem sizes (its multithreading threshold probably too small). Both SYR and SYRK interfaces lack a multithreading threshold completely, might be helpful to know how many threads MKL uses there. (And I assume that the main problem lies in the multithreading rather than the cpu-specific microkernels - that is, when you write "slow on Ryzen" you meant "slow, and btw I'm on Ryzen" rather than a defect on that platform alone ?)

**Comment by david-cortes** [comment]
Just tested on an old Pentium N3540. It has the same slow down in the SYR example above (~105s with OpenBLAS vs ~8s with MKL).

**Comment by david-cortes** [comment]
I'm also not so sure that the issue is with the multi-threading threshold, since setting `OMP_NUM_THREADS` and `OPENBLAS_NUM_THREADS` to 1 while parallelizing from an outer loop would still make it comparatively slower than MKL and Armadillo. As far as I understand, with those options set to one it should not use any multi-threading regardless of the threshold.

**Comment by david-cortes** [comment]
BTW this naive SYR (only for incx=1, upper, and row-major) achieves similar performance as MKL when compiled at O3.
```c
double temp;
double *restrict Arow;
 for (int i = 0; i < N; i++) {
    temp = alpha*X[i];
    Arow = A + (size_t)i*lda;
    for (int j = i; j < N; j++)
        Arow[j] += temp*X[j];
}
```

I imagine it might be reasonable to have a threshold before which it would only use that simple axpy-based loop (not sure what else it might be using right now).

**Comment by martin-frbg** [comment]
It is using an AXPY loop already (with optional handling of incx !=1 etc)

**Comment by brada4** [comment]
```
       liblapack.so.3 => /lib/x86_64-linux-gnu/liblapack.so.3 (0x00007f59dfee8000)
        libblas.so.3 => /lib/x86_64-linux-gnu/libblas.so.3 (0x00007f59dfe83000)
        libopenblas.so.0 => /lib/x86_64-linux-gnu/libopenblas.so.0 (0x00007f59dd377000)
```
Please run
`update-alternatives --config libblas.so.3` and `liblapack.so.3` and select same openblas. Otherwise it is not clear if your software build uses functions from OpenBLAS XOR NetLIB LAPACK , or worse a mix of both.

**Comment by david-cortes** [comment]
But the SYR loops in Jupyter were compiled using either `-lopenblas` or `-l:.. -L... -Wl,-rpath=...`. The Python build system doesn't automatically add BLAS/LAPACK linkage, unlike R.

However, I now suspect the issue might have something to do with either the Jupyter runtime or the Cython build system. I've been trying to reproduce the example SYR loop above in C++ without Python, and I do get the same timing for MKL (~0.7s), but for OpenBLAS I get wildly different results (~13s in Jupyter, ~1.8s in C++, ~1.4s in C++ with environment variables setting threads).

Any ideas of what else could Jupyter/Python/Cython be doing that affects performance? Or any ideas of what else to look at for debugging?

*************
This is the Jupyter example from before, redone in C++:
```cpp
#include <iostream>
#include <random>
#include <algorithm>
#include <cstdlib>
#include <vector>
#include <chrono>
#include "cblas.h"

void run_code
(
    const int k,
    const double *__restrict X_in,
    double *__restrict X_out,
    const int *indices, const int n,
    int nthreads
)
{
    #pragma omp parallel for schedule(static) num_threads(nthreads)
    for (int ix = 0; ix < n; ix++)
    {
        for (int repeat = 0; repeat < 1000; repeat++)
            cblas_dsyr(CblasRowMajor, CblasUpper, k,
                       1., X_in + indices[ix]*k, 1,
                       X_out + indices[ix]*k*k, k);
    }
}

int main(int arg, char **argv)
{
    int k = 50;
    int m = 10000;
    int nthreads = std::atoi(argv[1]);
    std::cout << "Using nthreads: " << nthreads << std::endl;
    

    std::vector<double> X_in(m*k);
    std::vector<double> X_out(m*k*k);
    std::vector<int> indices(m);

    std::default_random_engine rng(1);
    std::iota(indices.begin(), indices.end(), 0);
    std::shuffle(indices.begin(), indices.end(), rng);

    std::normal_distribution<double> rnorm{0,1};
    for (auto &el : X_in) el = rnorm(rng);

    auto tbegin = std::chrono::high_resolution_clock::now();

    run_code(
        k,
        X_in.data(),
        X_out.data(),
        indices.data(), m,
        nthreads
    );

    auto tend = std::chrono::high_resolution_clock::now();
    double elapsed_time_ms = std::chrono::duration<double, std::milli>(tend-tbegin).count();
    std::cout << "Time(s): " << (elapsed_time_ms / 1000.) << std::endl;

    /* ensure the compiler doesn't optimize out everything */
    double s = std::accumulate(X_out.begin(), X_out.end(), 0.);
    std::cout << "Sum is: " << s << std::endl;
    return 0;
}

```

I compile it like this:
```
g++ syr.cpp -O3 -std=c++11 -fopenmp -lopenblas
```

`ldd` shows:
```
david@debian:~/c_quicktest$ ldd a.out 
        linux-vdso.so.1 (0x00007ffcc0fbf000)
        libopenblas.so.0 => /lib/x86_64-linux-gnu/libopenblas.so.0 (0x00007f5eb3277000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f5eb30aa000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f5eb2f66000)
        libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f5eb2f26000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f5eb2f0c000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5eb2d47000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f5eb2d23000)
        libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007f5eb2a6d000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f5eb559b000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f5eb2a67000)
        libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f5eb2a1e000)
```

Then I run like this:
```
./a.out 16
```
(where 16 is the number of threads)

or like this:
```
OMP_NUM_THREADS=1 OPENBLAS_NUM_THREADS=1 ./a.out 16
```

Results:

* With MKL: ~0.7s
* With OpenBLAS: ~1.8s
* With OpenBLAS + env variables: ~1.4s

**Comment by martin-frbg** [comment]
The numpy in your python/cython example may depend on yet another libblas/liblapack. Other than that, no idea as not really familiar with jupyter.

**Comment by brada4** [comment]
@david-cortes could you measure with 
`perf record ./a.out 16`
`perf report`
Just copy like top 10 rows at least, or more if needed rows above and including  first occurrence of `libopenblas.so.0`
It is many things that could mess up things there.

**Comment by brada4** [comment]
I would criticise this:
`#pragma omp parallel for schedule(static) num_threads(nthreads)`
What starts with `for` is just redundant defaults of GOMP library.

**Comment by martin-frbg** [comment]
Most time spent in daxpy_k, daxpy_kernel_8 and blas_memory_alloc so no real surprises.

**Comment by MigMuc** [comment]
I tried the C++ code from above on Ryzen 3700X and OpenBLAS compiled like in Arch Linux except with USE_TLS=0 and have the following timings:
```
[xxx@xxx test_dsyr]$ OMP_NUM_THREADS=1 OPENBLAS_NUM_THREADS=1 ./a.out 8
Using nthreads: 8
Time(s): 0.874888
Sum is: 5.01048e+08
[xxx@xxx test_dsyr]$ OMP_NUM_THREADS=1 OPENBLAS_NUM_THREADS=1 ./a.out 16
Using nthreads: 16
Time(s): 0.81992
Sum is: 5.01048e+08```

**Comment by david-cortes** [comment]
Ran `perf record` under the setting that was running slow in Python. Surprisingly, this line came on top:
```
  69.59%  python   [kernel.kallsyms]                                  [k] native_queued_spin_lock_slowpath
```
And these are the sub-details (no idea what they mean):
```
 52.57 │       mov       0x8(%rdx),%eax                                                                 ▒
 37.98 │       test      %eax,%eax                                                                      ▒
  2.06 │     ↑ je        19d 
```

No idea how that happened or if it's anyhow related to OpenBLAS.

I also just realized that, despite the linking arguments, depending on the numpy/scipy version, as soon as it imports numpy/scipy it will load the CBLAS functions, and the code will actually run even without linking to OpenBLAs or MKL - but only in some versions, as I have different python environments and such a thing did not happen in all of them.

**Comment by martin-frbg** [comment]
You could try running the numpy show_config() , or `ldd` on one of the numpy libraries in your python site-config path, e.g. `ldd multiarray.so` (and similarly for scipy) to find out what they bring in. The spinlock thing does not appear in my traces of your C++ reproducer so may be something else (e.g. filesystem operation) waiting on a lock.

**Comment by david-cortes** [comment]
Finally managed to find the issue. Thing is, the NumPy there was loading its own OpenBLAS, which was the `pthreads` variant.

I just built OpenBLAS from source **without** `USE_OPENMP`, and re-compiled and re-ran the earlier C++ code. With the `pthreads` variant, it took ~9s to run, despite having `OMP_NUM_THREADS=1` and `OPENBLAS_NUM_THREADS=1`.

Even adding a call to `openblas_set_num_threads(1);` doesn't decrease the running time.

**Comment by martin-frbg** [comment]
Yes, number of threads plays very little role in this example. The pthreads build has a lot more locks than the OpenMP one just to make it thread-safe - obviously OpenMP manages to deal with concurrency issues much more gracefully and there may be room for improvement. OpenBLAS (and GotoBLAS before it) was totally not thread-safe until four years ago...

**Comment by david-cortes** [comment]
Update: building with USE_TLS=1 decreases running time from ~9s to ~5.8s in the C++ example.

Using `libiomp` instead of `libgomp` reduces the time to ~2.9s.

**Comment by martin-frbg** [comment]
Absolute baseline would be around 1s for USE_THREAD=0 (and absolutely no locking) with your OpenMP code - all the while screaming "bad memory unallocation" as the most obvious sign of races. I suspect the most promising option would be to create a shortcut path for small cases (and incx==1) much like the "direct sgemm" prototype i.e. call the AXPY kernel directly from the interface without bothering to allocate the buffer.

**Comment by brada4** [comment]
Let me translate, just make likely branch without allocation...

**Comment by martin-frbg** [comment]
getting ~0.7 sec for the SYR testcase now with the "fast path" on Ryzen5 4600H (12 threads)

**Comment by david-cortes** [comment]
Great! Tried the latest pthreads version from master and it's now down to ~2.2s for me, with both `libgomp` and `libiomp`, with `USE_TLS=1` and without it (vs. ~2.3s calling `cblas_daxpy` manually with openblas-pthreads and ~1s calling daxpy with openblas-openmp).

Although still, `gcc` somehow manages to make a much faster version (~0.55s) out of the naive loop when compiled with `-O3` and `-march=native`. Perhaps it uses a different loop type?

**Comment by martin-frbg** [comment]
Hmm.. it is a naive loop now, not sure why your build does not get faster than 2.2s. Are you sure you pulled from current develop ?

**Comment by david-cortes** [comment]
Yes, it is the `develop` branch, `openblas_config.h` shows:
```
#define OPENBLAS_VERSION "OpenBLAS 0.3.15.dev"
```
